components/eam/src/physics/cam/zm_conv.F90

```diff
@@ -886,8 +1544,152 @@ subroutine zm_convr(lchnk   ,ncol    , &
+           microp_st%accgirm(ideep(i),k)= loc_microp_st%accgirm(i,k)
+           microp_st%accgrim(ideep(i),k)= loc_microp_st%accgrim(i,k)
+           microp_st%accgrsm(ideep(i),k)= loc_microp_st%accgrsm(i,k)
+
+           microp_st%accgsln(ideep(i),k)= loc_microp_st%accgsln(i,k)
+           microp_st%accgsrn(ideep(i),k)= loc_microp_st%accgsrn(i,k)
+           microp_st%accgirn(ideep(i),k)= loc_microp_st%accgirn(i,k)
+
+           microp_st%accsrim(ideep(i),k)= loc_microp_st%accsrim(i,k)
+           microp_st%acciglm(ideep(i),k)= loc_microp_st%acciglm(i,k)
+           microp_st%accigrm(ideep(i),k)= loc_microp_st%accigrm(i,k)
+           microp_st%accsirm(ideep(i),k)= loc_microp_st%accsirm(i,k)
+
+           microp_st%accigln(ideep(i),k)= loc_microp_st%accigln(i,k)
+           microp_st%accigrn(ideep(i),k)= loc_microp_st%accigrn(i,k)
+           microp_st%accsirn(ideep(i),k)= loc_microp_st%accsirn(i,k)
+           microp_st%accgln(ideep(i),k) = loc_microp_st%accgln(i,k)
+           microp_st%accgrn(ideep(i),k) = loc_microp_st%accgrn(i,k)
+
+           microp_st%accilm(ideep(i),k) = loc_microp_st%accilm(i,k)
+           microp_st%acciln(ideep(i),k) = loc_microp_st%acciln(i,k)
+
+           microp_st%fallrm(ideep(i),k) = loc_microp_st%fallrm(i,k)
+           microp_st%fallsm(ideep(i),k) = loc_microp_st%fallsm(i,k)
+           microp_st%fallgm(ideep(i),k) = loc_microp_st%fallgm(i,k)
+           microp_st%fallrn(ideep(i),k) = loc_microp_st%fallrn(i,k)
+           microp_st%fallsn(ideep(i),k) = loc_microp_st%fallsn(i,k)
+           microp_st%fallgn(ideep(i),k) = loc_microp_st%fallgn(i,k)
+           microp_st%fhmrm (ideep(i),k) = loc_microp_st%fhmrm (i,k)
+         end if
       end do
    end do
+
+   if (zm_microp) then
+     do k = msg + 1,pver
+       do i = 1,ncol
+
+         !Interpolate variable from interface to mid-layer and convert it from units of "kg/kg" to "g/m3"
+
+         if(k.lt.pver) then
+            microp_st%qice (i,k) = 0.5_r8*(microp_st%qice(i,k)+microp_st%qice(i,k+1))
+            microp_st%qliq (i,k) = 0.5_r8*(microp_st%qliq(i,k)+microp_st%qliq(i,k+1))
+            microp_st%qrain (i,k) = 0.5_r8*(microp_st%qrain(i,k)+microp_st%qrain(i,k+1))
+            microp_st%qsnow (i,k) = 0.5_r8*(microp_st%qsnow(i,k)+microp_st%qsnow(i,k+1))
+            microp_st%qgraupel(i,k) = 0.5_r8*(microp_st%qgraupel(i,k)+microp_st%qgraupel(i,k+1))
+            microp_st%qni (i,k) = 0.5_r8*(microp_st%qni(i,k)+microp_st%qni(i,k+1))
+            microp_st%qnl (i,k) = 0.5_r8*(microp_st%qnl(i,k)+microp_st%qnl(i,k+1))
+            microp_st%qnr (i,k) = 0.5_r8*(microp_st%qnr(i,k)+microp_st%qnr(i,k+1))
+            microp_st%qns (i,k) = 0.5_r8*(microp_st%qns(i,k)+microp_st%qns(i,k+1))
+            microp_st%qng (i,k) = 0.5_r8*(microp_st%qng(i,k)+microp_st%qng(i,k+1))
+            microp_st%wu(i,k)   = 0.5_r8*(microp_st%wu(i,k)+microp_st%wu(i,k+1))
+         end if
+
+         if (t(i,k).gt. 273.15_r8 .and. t(i,k-1).le.273.15_r8) then
```

Comment:
Having "magic numbers" down in the code like this is one of my pet peeves. We should really be using the "shr" values for things like this, or at the very least put parameters like this at the top of the code as module level constant parameters, or at the top of the routine, so they are all in one place and we avoid using slightly different values in different parts of the code.

Reply:
273.15 is replaced by physical constant "tmelt".